### DISCUSSION TITLE
[feat] Add `ensureError` function

### DISCUSSION THREAD

**Discussion by stefaanv** [radashi-org/radashi#119] [Ideas]
As explained [here](https://medium.com/with-orus/93a9cbdf1af5), error handling in Typescript can be tedious because the catch returns `unknown` type.
I've been using the `ensureError` function every since I read the article, it could be a usefull addition to `radashi`.
```ts
function ensureError(value: unknown): Error {
  if (value instanceof Error) return value

  let stringified = '[Unable to stringify the thrown value]'
  try {
    stringified = JSON.stringify(value)
  } catch {}

  const error = new Error(`This value was thrown as is, not through an Error: ${stringified}`)
  return error
}
```

**Comment by aleclarson** [comment]
Hmm, I believe best practice is to only throw `Error` objects.

Of course, the place where this comes in handy is when you don't have control of what's being thrown. Or maybe you prefer a defensive coding style (“always assume the worst”). Although, I'd say there is such a thing as being “too defensive” and the proposed function may be crossing that line. In most cases, it's best to fix/fork the upstream package, instead of catering your error handling logic to its unadvisable behavior, and I don't want Radashi to encourage that.

**Comment by stefaanv** [comment]
@aleclarson 
Of course.  The problem is that - with the `strict` and `useUnknownInCatchVariables` compiler options turned ON (which they should be) - typescript assumes the error type to be of type `unknown` even when an `Error` is thrown, so it won't let you access the `message` property.  One can indicate the type between the catch brackets e.g. `catch (error: any)` but there Typescript only allows `unknown` and `any` types.  Setting the type to `any` opens the door to runtime errors in the `catch` block.

One option is to typecheck the error, this will mute Typescript
```ts
function throwing() {
  throw new Error('Just a error')
}

function test() {
  try {
    throwing()
  } catch (error) {
    if (error instanceof Error) {
      console.log(error)
    }
  }
}
```

The proposed function solves this problem in just one line of code without a type check.  
It's one line that has to be in every catch block and that shouldn't be necessary, but it's there because Javascript allows to throw non-Errors and because Typescript is (overly?) protective about that.
```ts
try {
  runFragileOperation()
} catch (err) {
  const error = ensureError(err)
  console.log(error.message)
}
```

I know the differences between the `type-checking` and `ensureError` options are very thin, but by including the latter in `radashi`, you give the developer the choice.

**Reply by aleclarson** [reply]
I think `error instanceof Error` is short enough to not necessitate a Radashi function. We're trying to avoid “syntax sugar” unless the task is error-prone to implement manually in plain TypeScript.

> […] by including `ensureError`, you give the developer the choice.

As an ideal, I believe we want to keep the number of choices as low as possible, therefore encouraging consistent/homogenous code.

**Comment by stefaanv** [comment]
OK, fair enough
